File: python/iforevents/integrations/_vendor.py
from __future__ import annotations

import importlib
from typing import Any, Dict


def load_vendor(module: str, extra: str) -> Any:
    try:
        return importlib.import_module(module)
    except ImportError as exc:
        raise ImportError(f'{module} is not installed; run: pip install "iforevents[{extra}]"') from exc


def scalarize(properties: Dict[str, Any]) -> Dict[str, Any]:
    """Vendors reject None and nested objects; keep scalars, stringify the rest."""
    out: Dict[str, Any] = {}
    for key, value in properties.items():
        if value is None:
            continue
        out[key] = value if isinstance(value, (str, int, float, bool)) else str(value)
    return out

File: js/packages/mixpanel/src/_vendor.ts
/** Loads an optional peer package at runtime; the message names what to install. */
export async function loadVendor<T>(specifier: string, pick?: (mod: Record<string, unknown>) => T): Promise<T> {
  let mod: Record<string, unknown>;
  try {
    mod = (await import(/* @vite-ignore */ specifier)) as Record<string, unknown>;
  } catch (cause) {
    throw new Error(`[iforevents] ${specifier} is not installed; run: npm install ${specifier}`, { cause });
  }
  return pick ? pick(mod) : ((mod.default ?? mod) as T);
}

/** Vendors reject nulls and nested objects; keep scalars, stringify the rest. */
export function scalarize(properties: Record<string, unknown>): Record<string, string | number | boolean> {
  const out: Record<string, string | number | boolean> = {};
  for (const [k, v] of Object.entries(properties)) {
    if (v === null || v === undefined) continue;
    out[k] = typeof v === "string" || typeof v === "number" || typeof v === "boolean" ? v : JSON.stringify(v);
  }
  return out;
}
